src/algebra/ring/basic.lean

```diff
@@ -149,23 +150,39 @@ protected def function.injective.non_unital_semiring
 See note [reducible non-instances]. -/
 @[reducible]
 protected def function.injective.non_assoc_semiring
-  {α : Type u} [non_assoc_semiring α] [has_one β]
+  {α : Type u} [non_assoc_semiring α]
+  {β : Type v} [has_mul β] [add_monoid_with_one β]
   (f : β → α) (hf : injective f) (zero : f 0 = 0) (one : f 1 = 1)
   (add : ∀ x y, f (x + y) = f x + f y) (mul : ∀ x y, f (x * y) = f x * f y) :
   non_assoc_semiring β :=
-{ .. hf.non_unital_non_assoc_semiring f zero add mul, .. hf.mul_one_class f one mul }
+{ .. ‹add_monoid_with_one β›,
+  .. hf.non_unital_non_assoc_semiring f zero add mul, .. hf.mul_one_class f one mul }
 
 /-- Pullback a `semiring` instance along an injective function.
 See note [reducible non-instances]. -/
 @[reducible]
 protected def function.injective.semiring
-  {α : Type u} [semiring α] [has_one β]
+  {α : Type u} [semiring α]
+  {β : Type v} [add_monoid_with_one β] [has_mul β]
   (f : β → α) (hf : injective f) (zero : f 0 = 0) (one : f 1 = 1)
   (add : ∀ x y, f (x + y) = f x + f y) (mul : ∀ x y, f (x * y) = f x * f y) :
   semiring β :=
-{ .. hf.monoid_with_zero f zero one mul, .. hf.add_comm_monoid f zero add,
+{ .. ‹add_monoid_with_one β›,
+  .. hf.monoid_with_zero f zero one mul, .. hf.add_comm_monoid f zero add,
   .. hf.distrib f add mul }
 
+/-- Pullback a `semiring` instance along an injective function.
+See note [reducible non-instances]. -/
+@[reducible]
+protected def function.injective.semiring'
```

Comment:
This causes the same problem that  #10152 introduced, which means that all the other `injective` lemmas for derived structures now do the wrong thing. I'm trying to fix this in #12126, but I'd rather not have the scope grow even larger while I'm working on it, due to this PR being merged...

Reply:
Ah, nevermind, I jumped to conclusions - this approach is less bad than making everything else assume the defaults.

Reply:
> Do you think the proper approach is to also add extra nat_cast and int_cast arguments to this function?

The `[add_monoid_with_one]` argument might be the right approach anyway, unless you want to a take a global `has_coe_t nat _` instance instead - with the `nsmul` / `npow` cases we have a `has_scalar` / `has_pow` typeclass to carry that extra argument.

Reply:
> I'm happy to wait for #12126.

Just as an update, I now have this passing CI, although there are a few small conflicts in the bors queue that will need resolving